File: backend/src/main/java/com/ecommerce/backend/service/ProductService.java
package com.ecommerce.backend.service;

import com.ecommerce.backend.dto.product.ProductDto;
import java.util.List;

public interface ProductService {
    List<ProductDto> getAllProducts();
    List<ProductDto> getAllActiveProducts();
    ProductDto getProductById(Long id);
    ProductDto createProduct(ProductDto productDto);
    ProductDto updateProduct(Long id, ProductDto productDto);
    void deleteProduct(Long id);
    List<ProductDto> searchProducts(String name, String description, Double minPrice, Double maxPrice);
}

File: frontend/src/app/modules/products/services/product.service.ts
import { Injectable } from '@angular/core';
import { ApiService } from '../../../core/services/api.service';
import { Observable } from 'rxjs';
import { Product } from '../models/product.interface';

@Injectable({
  providedIn: 'root'
})
export class ProductService {
  constructor(private api: ApiService) {}

  getProducts(): Observable<Product[]> {
    return this.api.get<Product[]>('products');
  }

  getActiveProducts(): Observable<Product[]> {
    return this.api.get<Product[]>('products/active');
  }

  getProductById(id: number): Observable<Product> {
    return this.api.get<Product>(`products/${id}`);
  }

  createProduct(product: Product): Observable<Product> {
    return this.api.post<Product>('products', product);
  }

  updateProduct(id: number, product: Product): Observable<Product> {
    return this.api.put<Product>(`products/${id}`, product);
  }

  deleteProduct(id: number): Observable<void> {
    return this.api.delete<void>(`products/${id}`);
  }

  searchProducts(term: string): Observable<Product[]> {
    return this.api.get<Product[]>('products/search', { q: term });
  }

  getTopSellingProducts(limit: number = 5): Observable<Product[]> {
    return this.api.get<Product[]>(`products/top-selling?limit=${limit}`);
  }
}